CMake/Autodownload.cmake

```diff
@@ -0,0 +1,84 @@
+function(get_deps LINK DEPS_NAME PACKAGE)
+
+  file(DOWNLOAD ${LINK}
+         "${CMAKE_BINARY_DIR}/deps/${PACKAGE}"
+          STATUS DOWNLOAD_STATUS_LIST LOG DOWNLOAD_LOG
+          SHOW_PROGRESS)
+    list(GET DOWNLOAD_STATUS_LIST 0 DOWNLOAD_STATUS)
+    if (DOWNLOAD_STATUS EQUAL 0)
+      if (NOT ${DEPS_NAME} MATCHES "stb")
+        execute_process(COMMAND ${CMAKE_COMMAND} -E
+            tar xf "${CMAKE_BINARY_DIR}/deps/${PACKAGE}"
+            WORKING_DIRECTORY "${CMAKE_BINARY_DIR}/deps/")
+
+        # Get the name of the directory.
+        file (GLOB DIRECTORIES RELATIVE "${CMAKE_BINARY_DIR}/deps/"
+            "${CMAKE_BINARY_DIR}/deps/${DEPS_NAME}*.*")
+        if(${DEPS_NAME} MATCHES "boost")
+          file (GLOB DIRECTORIES RELATIVE "${CMAKE_BINARY_DIR}/deps/"
+              "${CMAKE_BINARY_DIR}/deps/${DEPS_NAME}*_*")
+        endif()
+        # list(FILTER) is not available on 3.5 or older, but try to keep
+        # configuring without filtering the list anyway (it might work if only
+        # the file ensmallen-latest.tar.gz is present.
+        if (${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.6.0")
+          list(FILTER DIRECTORIES EXCLUDE REGEX ".*\.tar\.gz")
+          list(FILTER DIRECTORIES EXCLUDE REGEX ".*\.tar\.xz")
+        endif ()
+        list(LENGTH DIRECTORIES DIRECTORIES_LEN)
+        message("Print directories: " ${DIRECTORIES})
+        if (DIRECTORIES_LEN GREATER 0)
+          if (${DEPS_NAME} MATCHES "armadillo")
+            list(GET DIRECTORIES 0 ARMADILLO_DIR)
+            set(ARMADILLO_INCLUDE_DIR "${CMAKE_BINARY_DIR}/deps/${ARMADILLO_DIR}/include" CACHE INTERNAL "")
+
+          elseif (${DEPS_NAME} MATCHES "ensmallen")
+            list(GET DIRECTORIES 0 ENSMALLEN_DIR)
+            set(ENSMALLEN_INCLUDE_DIR "${CMAKE_BINARY_DIR}/deps/${ENSMALLEN_DIR}/include" CACHE INTERNAL "")
+
+          elseif (${DEPS_NAME} MATCHES "cereal")
+            list(GET DIRECTORIES 0 CEREAL_DIR)
+            set(CEREAL_INCLUDE_DIR "${CMAKE_BINARY_DIR}/deps/${CEREAL_DIR}/include" CACHE INTERNAL "")
+
+          elseif(${DEPS_NAME} MATCHES "boost")
+            list(GET DIRECTORIES 0 Boost_DIR)
+            set(Boost_INCLUDE_DIR "${CMAKE_BINARY_DIR}/deps/${Boost_DIR}/" CACHE INTERNAL "")
+          
+          elseif(${DEPS_NAME} MATCHES "OpenBLAS")
+            list(GET DIRECTORIES 0 OPENBLAS_DIR)
+            execute_process(COMMAND make TARGET=ARMV8 BINARY=64 HOSTCC=gcc CC=${CMAKE_C_COMPILER} FC=${CMAKE_FORTRAN_COMPILER} NO_SHARED=1
+                            WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/deps/${OPENBLAS_DIR})
```

Comment:
This may not work on Windows... maybe we should check to see `if (MSVC)` or something like that and disallow auto-downloading if so?  (I suppose technically a MinGW environment might have `make` on the path.  But MSVC probably won't, at least.)